### ISSUE TITLE
训练量化接口转离线，测试时卡死以及可能存在内存泄露？

### ISSUE THREAD

**Issue by phecda-xu** [alibaba/MNN#1468]
# 平台(如果交叉编译请再附上交叉编译目标平台):  ubuntu 16.04，24线程CPU, 32G内存；
pip安装,版本为 1.1.4

参照 [quant_aware_training.py](https://github.com/alibaba/MNN/blob/3788948b661ca98e6bb5196af76e6a49f39589f8/pymnn/examples/MNNTrain/quantization_aware_training/quant_aware_training.py) 改写

```
def test_func(net, test_dataloader):
    net.train(False)
    test_dataloader.reset()

    correct = 0
    total = 0
    for i in range(test_dataloader.iter_number):
        example = test_dataloader.next()
        input_data = example[0]
        output_target = example[1]
        data = input_data[0]  # which input, model may have more than one inputs
        label = output_target[0]  # also, model may have more than one outputs
        predict = net.forward(data)
        predict.read()

```

测试函数在执行到 `net(data)`时卡死；

此外，在训练函数执行时，出现了内存随着 迭代次数增加缓慢增长的情况，知道最后程序被kill掉。

```
def train_func(net, train_dataloader, opt, num_classes):
    net.train(True)
    train_dataloader.reset()

    t0 = time.time()
    # for i in range(train_dataloader.iter_number):
    for i in range(100): # actually, in our full experiment, we only need 3K images using ILSVRC2012 training dataset
        example = train_dataloader.next()
        input_data = example[0]
        output_target = example[1]
        data = input_data[0]  # which input, model may have more than one inputs
        label = output_target[0]  # also, model may have more than one outputs
        predict = net.forward(data)
        predict.read()
```
尝试过注释掉`predict = net.forward(data)` ，只加载数据时内存是正常的，一旦执行forward 推理就再次出现内存不够的情况，疑似内存泄露。

求分析原因和确认情况，谢谢！

**Comment by yyfcc17** [comment]
现在可以使用这个工具，https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant
batchsize减小点试试32，或64

**Comment by phecda-xu** [comment]
> 现在可以使用这个工具，https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant
> batchsize减小点试试32，或64

谢谢，我尝试后发现应该是 遗漏了 `opt.step(0.0,[])` 这个操作，加上后两个情况都解决了；


参照 https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant 的描述，`需要使用--forTraining 转换的mnn模型`，下面是我在实验观察到的一些情况：

我是用pip直接安装的MNN，带了一个mnnconvert工具，我之前是用的这个进行模型转换（onnx -> mnn）。

但是mnnconvert 不支持 --forTraining 设置，支持设置如下：
```
usage: mnnconvert [-h] -f {TF,CAFFE,ONNX,TFLITE,MNN} --modelFile MODELFILE
                  [--prototxt PROTOTXT] --MNNModel MNNMODEL --bizCode BIZCODE
                  [--fp16 FP16] [--weightQuantBits WEIGHTQUANTBITS]
                  [--weightQuantAsymmetric WEIGHTQUANTASYMMETRIC]
                  [--compressionParamsFile COMPRESSIONPARAMSFILE]
```
参照语雀文档，我尝试了单独在MNN中进行了编译，生成了 MNNCovert 工具，然后用它来转模型格式（onnx -> mnn）

用 MNNCovert 转换的工具生成的模型，再去执行原来的训练转离线脚本。执行顺利，没有遇到报错。

上述两个工具转换的模型有区别，

我用下面的代码加载转换后的模型进行输入输出测试，
```
interpreter = MNN.Interpreter(MNN_modelt)
session = interpreter.createSession()
input_tensor = interpreter.getSessionInput(session)
input_shape = input_tensor.getShape()
print("input_shape = {}".format(input_shape))

output_tensor = interpreter.getSessionOutput(session)
output_shape = output_tensor.getShape()
print("output_shape = {}".format(output_shape))

tmp_input = MNN.Tensor((1, 1, 161, 132), MNN.Halide_Type_Float,
                       input_data.detach().numpy(),
                       MNN.Tensor_DimensionType_Caffe)
input_tensor.copyFrom(tmp_input)
interpreter.runSession(session)
output_tensor = interpreter.getSessionOutput(session)

# constuct a tmp tensor and copy/convert in case output_tensor is nc4hw4
tmp_output = MNN.Tensor((110, 327), MNN.Halide_Type_Float, np.ones([110, 327]).astype(np.float32),
                        MNN.Tensor_DimensionType_Caffe)
output_tensor.copyToHostTensor(tmp_output)
mnn_output = tmp_output.getData()
mnn_output = np.array(mnn_output).reshape([110, 327])
```
- 不加 --forTraining 设置，mnnconvert 与 MNNConvert 表现一致
-  加 --forTraining 设置， mnnconvert 无对应的设置，无法转换模型； MNNConvert  可以转换得到模型，但是输出的结果值全为1
且出现以下的提示信息：

```
Create exection error : 267
...
Can't run session because not resized
``` 
即： MNNConvert + --forTraining 转换的mnn模型，使用 interpreter.createSession() 方法来进行模型推理存在未知的错误。


我看到模型转换参数说明里面提到，--forTraining 时为了保存训练相关算子如BN/Dropout， 也就是说如果模型中存在这些结构，需要保留，然后在训练量化中参与训练。如果onnx模型不包含这些算子，是不是就不需要--forTraining设置，使用训练转离线量化也依然有效果？

**Comment by yjiangling** [comment]
> > 现在可以使用这个工具，https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant
> > batchsize减小点试试32，或64
> 
> 谢谢，我尝试后发现应该是 遗漏了 `opt.step(0.0,[])` 这个操作，加上后两个情况都解决了；
> 
> 参照 https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant 的描述，`需要使用--forTraining 转换的mnn模型`，下面是我在实验观察到的一些情况：
> 
> 我是用pip直接安装的MNN，带了一个mnnconvert工具，我之前是用的这个进行模型转换（onnx -> mnn）。
> 
> 但是mnnconvert 不支持 --forTraining 设置，支持设置如下：
> 
> ```
> usage: mnnconvert [-h] -f {TF,CAFFE,ONNX,TFLITE,MNN} --modelFile MODELFILE
>                   [--prototxt PROTOTXT] --MNNModel MNNMODEL --bizCode BIZCODE
>                   [--fp16 FP16] [--weightQuantBits WEIGHTQUANTBITS]
>                   [--weightQuantAsymmetric WEIGHTQUANTASYMMETRIC]
>                   [--compressionParamsFile COMPRESSIONPARAMSFILE]
> ```
> 
> 参照语雀文档，我尝试了单独在MNN中进行了编译，生成了 MNNCovert 工具，然后用它来转模型格式（onnx -> mnn）
> 
> 用 MNNCovert 转换的工具生成的模型，再去执行原来的训练转离线脚本。执行顺利，没有遇到报错。
> 
> 上述两个工具转换的模型有区别，
> 
> 我用下面的代码加载转换后的模型进行输入输出测试，
> 
> ```
> interpreter = MNN.Interpreter(MNN_modelt)
> session = interpreter.createSession()
> input_tensor = interpreter.getSessionInput(session)
> input_shape = input_tensor.getShape()
> print("input_shape = {}".format(input_shape))
> 
> output_tensor = interpreter.getSessionOutput(session)
> output_shape = output_tensor.getShape()
> print("output_shape = {}".format(output_shape))
> 
> tmp_input = MNN.Tensor((1, 1, 161, 132), MNN.Halide_Type_Float,
>                        input_data.detach().numpy(),
>                        MNN.Tensor_DimensionType_Caffe)
> input_tensor.copyFrom(tmp_input)
> interpreter.runSession(session)
> output_tensor = interpreter.getSessionOutput(session)
> 
> # constuct a tmp tensor and copy/convert in case output_tensor is nc4hw4
> tmp_output = MNN.Tensor((110, 327), MNN.Halide_Type_Float, np.ones([110, 327]).astype(np.float32),
>                         MNN.Tensor_DimensionType_Caffe)
> output_tensor.copyToHostTensor(tmp_output)
> mnn_output = tmp_output.getData()
> mnn_output = np.array(mnn_output).reshape([110, 327])
> ```
> 
> * 不加 --forTraining 设置，mnnconvert 与 MNNConvert 表现一致
> * 加 --forTraining 设置， mnnconvert 无对应的设置，无法转换模型； MNNConvert  可以转换得到模型，但是输出的结果值全为1
>   且出现以下的提示信息：
> 
> ```
> Create exection error : 267
> ...
> Can't run session because not resized
> ```
> 
> 即： MNNConvert + --forTraining 转换的mnn模型，使用 interpreter.createSession() 方法来进行模型推理存在未知的错误。
> 
> 我看到模型转换参数说明里面提到，--forTraining 时为了保存训练相关算子如BN/Dropout， 也就是说如果模型中存在这些结构，需要保留，然后在训练量化中参与训练。如果onnx模型不包含这些算子，是不是就不需要--forTraining设置，使用训练转离线量化也依然有效果？

哈喽，请教一下你这个最后是怎么解决的，我在模型转换时没有加上--forTraining参数设置，仍然报错Can't run session because not resized

**Comment by phecda-xu** [comment]
> > > 现在可以使用这个工具，https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant
> > > batchsize减小点试试32，或64
> > 
> > 
> > 谢谢，我尝试后发现应该是 遗漏了 `opt.step(0.0,[])` 这个操作，加上后两个情况都解决了；
> > 参照 https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant 的描述，`需要使用--forTraining 转换的mnn模型`，下面是我在实验观察到的一些情况：
> > 我是用pip直接安装的MNN，带了一个mnnconvert工具，我之前是用的这个进行模型转换（onnx -> mnn）。
> > 但是mnnconvert 不支持 --forTraining 设置，支持设置如下：
> > ```
> > usage: mnnconvert [-h] -f {TF,CAFFE,ONNX,TFLITE,MNN} --modelFile MODELFILE
> >                   [--prototxt PROTOTXT] --MNNModel MNNMODEL --bizCode BIZCODE
> >                   [--fp16 FP16] [--weightQuantBits WEIGHTQUANTBITS]
> >                   [--weightQuantAsymmetric WEIGHTQUANTASYMMETRIC]
> >                   [--compressionParamsFile COMPRESSIONPARAMSFILE]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 参照语雀文档，我尝试了单独在MNN中进行了编译，生成了 MNNCovert 工具，然后用它来转模型格式（onnx -> mnn）
> > 用 MNNCovert 转换的工具生成的模型，再去执行原来的训练转离线脚本。执行顺利，没有遇到报错。
> > 上述两个工具转换的模型有区别，
> > 我用下面的代码加载转换后的模型进行输入输出测试，
> > ```
> > interpreter = MNN.Interpreter(MNN_modelt)
> > session = interpreter.createSession()
> > input_tensor = interpreter.getSessionInput(session)
> > input_shape = input_tensor.getShape()
> > print("input_shape = {}".format(input_shape))
> > 
> > output_tensor = interpreter.getSessionOutput(session)
> > output_shape = output_tensor.getShape()
> > print("output_shape = {}".format(output_shape))
> > 
> > tmp_input = MNN.Tensor((1, 1, 161, 132), MNN.Halide_Type_Float,
> >                        input_data.detach().numpy(),
> >                        MNN.Tensor_DimensionType_Caffe)
> > input_tensor.copyFrom(tmp_input)
> > interpreter.runSession(session)
> > output_tensor = interpreter.getSessionOutput(session)
> > 
> > # constuct a tmp tensor and copy/convert in case output_tensor is nc4hw4
> > tmp_output = MNN.Tensor((110, 327), MNN.Halide_Type_Float, np.ones([110, 327]).astype(np.float32),
> >                         MNN.Tensor_DimensionType_Caffe)
> > output_tensor.copyToHostTensor(tmp_output)
> > mnn_output = tmp_output.getData()
> > mnn_output = np.array(mnn_output).reshape([110, 327])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * 不加 --forTraining 设置，mnnconvert 与 MNNConvert 表现一致
> > * 加 --forTraining 设置， mnnconvert 无对应的设置，无法转换模型； MNNConvert  可以转换得到模型，但是输出的结果值全为1
> >   且出现以下的提示信息：
> > 
> > ```
> > Create exection error : 267
> > ...
> > Can't run session because not resized
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 即： MNNConvert + --forTraining 转换的mnn模型，使用 interpreter.createSession() 方法来进行模型推理存在未知的错误。
> > 我看到模型转换参数说明里面提到，--forTraining 时为了保存训练相关算子如BN/Dropout， 也就是说如果模型中存在这些结构，需要保留，然后在训练量化中参与训练。如果onnx模型不包含这些算子，是不是就不需要--forTraining设置，使用训练转离线量化也依然有效果？
> 
> 哈喽，请教一下你这个最后是怎么解决的，我在模型转换时没有加上--forTraining参数设置，仍然报错Can't run session because not resized

版本更新到1.1.4以上，目前只有加--forTraining 设置转的模型会出这个错，其它的都是正常的，出错的模型经过训练接口转离线量化后的模型也是正常的。 目前还没有找到导致 `Can't run session because not resized` 这个错误的原因。建议直接去群里问问。

**Comment by yjiangling** [comment]
> > > > 现在可以使用这个工具，https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant
> > > > batchsize减小点试试32，或64
> > > 
> > > 
> > > 谢谢，我尝试后发现应该是 遗漏了 `opt.step(0.0,[])` 这个操作，加上后两个情况都解决了；
> > > 参照 https://github.com/alibaba/MNN/tree/master/tools/MNNPythonOfflineQuant 的描述，`需要使用--forTraining 转换的mnn模型`，下面是我在实验观察到的一些情况：
> > > 我是用pip直接安装的MNN，带了一个mnnconvert工具，我之前是用的这个进行模型转换（onnx -> mnn）。
> > > 但是mnnconvert 不支持 --forTraining 设置，支持设置如下：
> > > ```
> > > usage: mnnconvert [-h] -f {TF,CAFFE,ONNX,TFLITE,MNN} --modelFile MODELFILE
> > >                   [--prototxt PROTOTXT] --MNNModel MNNMODEL --bizCode BIZCODE
> > >                   [--fp16 FP16] [--weightQuantBits WEIGHTQUANTBITS]
> > >                   [--weightQuantAsymmetric WEIGHTQUANTASYMMETRIC]
> > >                   [--compressionParamsFile COMPRESSIONPARAMSFILE]
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 参照语雀文档，我尝试了单独在MNN中进行了编译，生成了 MNNCovert 工具，然后用它来转模型格式（onnx -> mnn）
> > > 用 MNNCovert 转换的工具生成的模型，再去执行原来的训练转离线脚本。执行顺利，没有遇到报错。
> > > 上述两个工具转换的模型有区别，
> > > 我用下面的代码加载转换后的模型进行输入输出测试，
> > > ```
> > > interpreter = MNN.Interpreter(MNN_modelt)
> > > session = interpreter.createSession()
> > > input_tensor = interpreter.getSessionInput(session)
> > > input_shape = input_tensor.getShape()
> > > print("input_shape = {}".format(input_shape))
> > > 
> > > output_tensor = interpreter.getSessionOutput(session)
> > > output_shape = output_tensor.getShape()
> > > print("output_shape = {}".format(output_shape))
> > > 
> > > tmp_input = MNN.Tensor((1, 1, 161, 132), MNN.Halide_Type_Float,
> > >                        input_data.detach().numpy(),
> > >                        MNN.Tensor_DimensionType_Caffe)
> > > input_tensor.copyFrom(tmp_input)
> > > interpreter.runSession(session)
> > > output_tensor = interpreter.getSessionOutput(session)
> > > 
> > > # constuct a tmp tensor and copy/convert in case output_tensor is nc4hw4
> > > tmp_output = MNN.Tensor((110, 327), MNN.Halide_Type_Float, np.ones([110, 327]).astype(np.float32),
> > >                         MNN.Tensor_DimensionType_Caffe)
> > > output_tensor.copyToHostTensor(tmp_output)
> > > mnn_output = tmp_output.getData()
> > > mnn_output = np.array(mnn_output).reshape([110, 327])
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 
> > > * 不加 --forTraining 设置，mnnconvert 与 MNNConvert 表现一致
> > > * 加 --forTraining 设置， mnnconvert 无对应的设置，无法转换模型； MNNConvert  可以转换得到模型，但是输出的结果值全为1
> > >   且出现以下的提示信息：
> > > 
> > > ```
> > > Create exection error : 267
> > > ...
> > > Can't run session because not resized
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 即： MNNConvert + --forTraining 转换的mnn模型，使用 interpreter.createSession() 方法来进行模型推理存在未知的错误。
> > > 我看到模型转换参数说明里面提到，--forTraining 时为了保存训练相关算子如BN/Dropout， 也就是说如果模型中存在这些结构，需要保留，然后在训练量化中参与训练。如果onnx模型不包含这些算子，是不是就不需要--forTraining设置，使用训练转离线量化也依然有效果？
> > 
> > 
> > 哈喽，请教一下你这个最后是怎么解决的，我在模型转换时没有加上--forTraining参数设置，仍然报错Can't run session because not resized
> 
> 版本更新到1.1.4以上，目前只有加--forTraining 设置转的模型会出这个错，其它的都是正常的，出错的模型经过训练接口转离线量化后的模型也是正常的。 目前还没有找到导致 `Can't run session because not resized` 这个错误的原因。建议直接去群里问问。

好的，感谢解答~
